ethcore/src/client.rs

```diff
@@ -243,85 +243,117 @@ impl Client {
 		self.block_queue.write().unwrap().flush();
 	}
 
+	fn build_last_hashes(&self, header: &Header) -> LastHashes {
+		let mut last_hashes = LastHashes::new();
+		last_hashes.resize(256, H256::new());
+		last_hashes[0] = header.parent_hash.clone();
+		let chain = self.chain.read().unwrap();
+		for i in 0..255 {
+			match chain.block_details(&last_hashes[i]) {
+				Some(details) => {
+					last_hashes[i + 1] = details.parent.clone();
+				},
+				None => break,
+			}
+		}
+		last_hashes
+	}
+
 	/// This is triggered by a message coming from a block queue when the block is ready for insertion
 	pub fn import_verified_blocks(&self, io: &IoChannel<NetSyncMessage>) -> usize {
-		let mut ret = 0;
-		let mut bad = HashSet::new();
+		let max_blocks_to_import = 128;
+
+		let mut imported = 0;
+		let mut good_blocks = Vec::with_capacity(max_blocks_to_import);
+		let mut bad_blocks = HashSet::new();
+		let engine = self.engine.deref().deref();
+
 		let _import_lock = self.import_lock.lock();
-		let blocks = self.block_queue.write().unwrap().drain(128);
-		let mut good_blocks = Vec::with_capacity(128);
+		let blocks = self.block_queue.write().unwrap().drain(max_blocks_to_import);
+
 		for block in blocks {
-			if bad.contains(&block.header.parent_hash) {
-				self.block_queue.write().unwrap().mark_as_bad(&block.header.hash());
-				bad.insert(block.header.hash());
+			let header = &block.header;
+			let header_hash = block.header.hash();
+			let bad_contains_parent = bad_blocks.contains(&header.parent_hash);
+
+			let mark_block_as_bad = || {
+				self.block_queue.write().unwrap().mark_as_bad(&header_hash);
```

Comment:
Is it really necessary to mark the blocks in queue in the loop? Could it be done like with `mark_as_good(&[H256])` ?